### PR TITLE
Replace deprecated torch.cuda.amp decorators for torch 2.4

### PR DESCRIPTION
**Pull request by loichuder**
#### Changes
Version 2.4.0 of `pytorch` deprecates device-specific autocast API: https://github.com/pytorch/pytorch/releases/tag/v2.4.0. Therefore, instructions such as `@torch.cuda.amp...` raise a  warning:

```
FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
```

This PR changes device-specific autocasts by their new syntax where the device is an argument.

#### Type of change
**Uuuh.... Not sure ?**
- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Comment by loichuder** [comment]
Precommit failures are fixed by #2966. Not sure how to fix the other failures though...

**Review by johnnv1 — COMMENTED** [review]
overall everything is fine, but could you find a solution to the typing problem? The other problems that fail on CI are not related to your patch and are fine

File: kornia/contrib/models/efficient_vit/nn/ops.py [hunk 1]
```diff
@@ -8,7 +8,17 @@
...
+from kornia.utils._compat import torch_version_ge
+
+if torch_version_ge(2, 4):
+    from functools import partial
+
+    from torch.amp import autocast as _autocast
+
+    autocast = partial(_autocast, "cuda")
+else:
+    from torch.cuda.amp import autocast
```
Comment (johnnv1) [inline kornia/contrib/models/efficient_vit/nn/ops.py:21]: maybe we should do these imports under the `kornia.utils._compat`, and import it from there? cc @edgarriba

**Comment by loichuder** [comment]
> overall everything is fine, but could you find a solution to the typing problem?

Well, it comes from the fact that `mypy` does not like conditional imports and there are apparently multiple solutions to this: https://github.com/python/mypy/issues/1153

I tried to do this with some finesse (https://github.com/python/mypy/issues/1153#issuecomment-1167676627) but it didn't work so I resorted to the ugly workaround with `TYPE_CHECKING`: https://github.com/python/mypy/issues/1153#issuecomment-1207333806

File: kornia/feature/lightglue.py [hunk 2]
```diff
@@ -36,12 +38,23 @@
 else:
     FLASH_AVAILABLE = False
 
+if TYPE_CHECKING:
```
Comment (johnnv1) [inline kornia/feature/lightglue.py:41]:
```suggestion
if TYPE_CHECKING: # TODO (@johnnv1): remove this branch when bump the pytorch CI to support torch 2.4
```

Comment (johnnv1) [reply kornia/feature/lightglue.py:41]: I'm going to try and get that chore up in the next hout

**Review by edgarriba — APPROVED** [review]

**Pull request merged**